File: apps/android/app/src/main/java/com/tonezen/app/domain/downloads/DownloadResumePolicy.kt
package com.tonezen.app.domain.downloads

object DownloadResumePolicy {
    const val COMPLETED_HISTORY_LIMIT = 200

    enum class ResumeAction {
        FULL_DOWNLOAD,
        RANGE_APPEND,
        RESTART,
    }

    fun resolveResumeAction(
        partFileLength: Long,
        bytesDownloaded: Long,
        totalBytes: Long?,
        rangeResponseCode: Int?,
    ): ResumeAction {
        if (partFileLength <= 0L && bytesDownloaded <= 0L) return ResumeAction.FULL_DOWNLOAD
        if (totalBytes != null && totalBytes > 0 && partFileLength > totalBytes) return ResumeAction.RESTART
        if (partFileLength != bytesDownloaded && bytesDownloaded > 0) {
            return if (partFileLength > bytesDownloaded) ResumeAction.RESTART else ResumeAction.RANGE_APPEND
        }
        return when (rangeResponseCode) {
            206 -> ResumeAction.RANGE_APPEND
            200, 416 -> ResumeAction.RESTART
            null -> if (partFileLength > 0L) ResumeAction.RANGE_APPEND else ResumeAction.FULL_DOWNLOAD
            else -> ResumeAction.RESTART
        }
    }

    fun progressFraction(bytesDownloaded: Long, totalBytes: Long?): Float? {
        if (totalBytes == null || totalBytes <= 0L) return null
        return (bytesDownloaded.toFloat() / totalBytes.toFloat()).coerceIn(0f, 1f)
    }
}

File: apps/desktop/src/shared/downloadResumePolicy.ts
export const COMPLETED_HISTORY_LIMIT = 200;

export type ResumeAction = "FULL_DOWNLOAD" | "RANGE_APPEND" | "RESTART";

export function resolveResumeAction(
  partFileLength: number,
  bytesDownloaded: number,
  totalBytes: number | null,
  rangeResponseCode: number | null,
): ResumeAction {
  if (partFileLength <= 0 && bytesDownloaded <= 0) return "FULL_DOWNLOAD";
  if (totalBytes != null && totalBytes > 0 && partFileLength > totalBytes) return "RESTART";
  if (partFileLength !== bytesDownloaded && bytesDownloaded > 0) {
    return partFileLength > bytesDownloaded ? "RESTART" : "RANGE_APPEND";
  }
  switch (rangeResponseCode) {
    case 206:
      return "RANGE_APPEND";
    case 200:
    case 416:
      return "RESTART";
    case null:
      return partFileLength > 0 ? "RANGE_APPEND" : "FULL_DOWNLOAD";
    default:
      return "RESTART";
  }
}

export function progressFraction(bytesDownloaded: number, totalBytes: number | null): number | null {
  if (totalBytes == null || totalBytes <= 0) return null;
  return Math.min(Math.max(bytesDownloaded / totalBytes, 0), 1);
}
